Add "prefiled" as public status option

bfreeds opened the pull request:
Part of the CRM update was to add a prefiled optinon.  We addressed this
in PR #79, however we missed this place the dcp_publicstatus options
are listed.

This PR fixes this and closes #111.

allthesignals commented:
Closes https://github.com/NYCPlanning/labs-zap-search/issues/424

allthesignals commented on routes/projects/index.js:
```diff
@@ -2,19 +2,78 @@ const express = require('express');
+
+/* GET /projects/download/shp */
+/* downloads a shapefile of projects that match the query params */
+router.get('/download/shp', async (req, res) => {
```
Why should this be above the most important part of this route definition? It should also be a separate route file, if we really want the route to be `/download/shp`.

I also expected this to be like a proper file extension:
`/projects.shp?q=whatever`

Can we clean this up so this file doesn't grow too large and unwieldy?

chriswhong replied:
Agree, we should clean this up and split it out to another file

chriswhong replied:
Added `/projects.{ext}`

allthesignals commented on routes/projects/index.js:
```diff
@@ -2,19 +2,78 @@ const express = require('express');
+      };
+    }),
+  };
+};
```
Can we put this into a separate utility file?

allthesignals commented on routes/projects/index.js:
```diff
@@ -2,19 +2,78 @@ const express = require('express');
+      .stream();
+    shapefile.pipe(res);
+  } catch (e) {
+    res.status(404).send({
```
404 is the wrong error code - this should be a 500 internal server error, not a 404 not found.

chriswhong replied:
👍

chriswhong replied:
fixed

allthesignals commented on routes/projects/index.js:
```diff
@@ -2,19 +2,78 @@ const express = require('express');
+
+/* GET /projects/download/geojson */
+/* downloads a geojson FeatureCollection of projects that match the query params */
+router.get('/download/geojson', async (req, res) => {
```
Same as comments about, I expected this to be:
`/projects.geojson?query=wahtever`.

This is a representation of the projects resource with a given extension.

I know it doesn't currently work this way with `/download.csv`, but maybe it should at least go in that route? I think we should use this issue to clean up the approach a bit and avoid making the index route any more complicated.

chriswhong replied:
+1, let's make all 3 download types the same

allthesignals commented on routes/projects/index.js:
```diff
@@ -2,19 +2,78 @@ const express = require('express');
+    res.setHeader('Content-type', 'application/json');
+    res.end(JSON.stringify(FeatureCollection));
+  } catch (e) {
+    res.status(404).send({
```
Same as above, this would be a 500 not a 404

chriswhong replied:
No longer relevant, index.js was simplified

allthesignals commented on routes/projects/index.js:
```diff
@@ -2,19 +2,78 @@ const express = require('express');
+    res.end(JSON.stringify(FeatureCollection));
+  } catch (e) {
+    res.status(404).send({
+      error: e.toString(),
```
FYI I think toString will actually eat up the stack trace and make it harder to see what's happening, not sure though

allthesignals commented on routes/projects/index.js:
```diff
@@ -104,4 +163,7 @@ router.get('/', async (req, res) => {
 });
 
+// moved to last so we can have a higher priority direct match
+router.use('/:id', require('./project'));
```
Can you comment & explain this a little more?

chriswhong replied:
If you have `projects/anything` it will match against `projects/id`.  I think it doesn't matter anymore, but this was here because I was originally trying to do `/projects/download.geojson` which was matching `projects/:id`

allthesignals commented on utils/build-projects-sql.js:
```diff
@@ -7,6 +7,7 @@ const getQueryFile = require('../utils/get-query-file');
 const listProjectsQuery = getQueryFile('/projects/index.sql');
 const paginateQuery = getQueryFile('/helpers/paginate.sql');
 const standardColumns = getQueryFile('/helpers/standard-projects-columns.sql');
+const shpColumns = getQueryFile('/helpers/shp-projects-columns.sql');
```
Why is this called `shpColumns`? Is there a special column for shapefile?

chriswhong replied:
It's a separate set of columns for the `SELECT`, which should include geometry and whatever we want in the shapefile/geojson attributes

chriswhong replied:
changed to `spatialColumns`

allthesignals commented on utils/build-projects-sql.js:
```diff
@@ -49,6 +50,7 @@ const buildProjectsSQL = (queryParams, type = 'filter') => {
   const blockQuery = block ? pgp.as.format("AND (blocks ilike '%$1:value%')", [block]) : '';
   const radiusQuery = distance_from_point[0] ? pgp.as.format('AND ST_Distance_Sphere(c.polygons, ST_MakePoint($1,$2)) <= $3', [...distance_from_point, ((radius_from_point / 5280) * 1609.34)]) : '';
   const paginate = generateDynamicQuery(paginateQuery, { itemsPerPage, offset: (page - 1) * itemsPerPage });
+  const hasPolygonsQuery = '';
```
`hasPolygonsQuery` implies a boolean, not a string. Can we name it so it's not implying the wrong type?

chriswhong replied:
+1

chriswhong replied:
This was removed

allthesignals commented on utils/build-projects-sql.js:
```diff
@@ -112,10 +116,35 @@ const buildProjectsSQL = (queryParams, type = 'filter') => {
+      radiusQuery,
+      hasPolygonsQuery: 'AND polygons IS NOT NULL',
+      paginate: '',
+    });
```
Hmmm how is this different from `(type === 'download')`?

chriswhong replied:
Not following, this is just to exclude projects that have no geometry from the spatial downloads (we don't want any rows with no geometries in our geojson or shapefile)

chriswhong replied:
this was removed

allthesignals commented on routes/projects/download.js:
```diff
@@ -3,70 +3,124 @@ const Json2csvTransform = require('json2csv').Transform;
+  res.setHeader('Content-disposition', `attachment; filename=projects.${filetype === 'shp' ? 'zip' : filetype}`);
+
+  // Flush the headers before we start pushing the CSV content
+  // res.flushHeaders();
```
Remove

allthesignals commented on routes/projects/download.js:
```diff
@@ -3,70 +3,124 @@ const Json2csvTransform = require('json2csv').Transform;
+  // res.flushHeaders();
+
+
+  if (filetype === 'csv') {
```
Wait, I don't follow the control flow here... if it's a csv request, do csv munging stuff, else check if it's `shp`, do shape stuff else return json. Is that right? Wouldn't it be easier to read if it were like this:

```pseudocode
if (csv) {
  // do csv stuff
}

if (shp) {
  // do shp stuff
}

if (json) {
  // do json stuff
}
```
Thoughts?

chriswhong replied:
I wrote it this way because shp and geojson share the same `getProjectsFeatureCollection` call, but they can just do it on their own... it's easier to read the way you suggest.

allthesignals commented on routes/projects/download.js:
```diff
@@ -3,70 +3,124 @@ const Json2csvTransform = require('json2csv').Transform;
+        });
+    } else {
+      res.status(500).send({
+        error: 'no data',
```
You seem to handle the "no data" case for CSV, but not for other types. How come? Why not just have this logic up top?

chriswhong replied:
no longer relevant

allthesignals commented on routes/projects/index.js:
```diff
@@ -71,7 +71,9 @@ router.get('/download/geojson', async (req, res) => {
-router.use('/download.csv', require('./download'));
+// router.use('.:extension', (req, res) => {
+//   res.send('O HAI')
+// })
```
nope

allthesignals commented on routes/projects/download.js:
```diff
@@ -3,70 +3,124 @@ const Json2csvTransform = require('json2csv').Transform;
+          );
+        })
+        .catch((error) => {
+          console.log('ERROR:', error); // eslint-disable-line
```
If there's an error, shouldn't it be a res.send error object?

chriswhong replied:
it is now

allthesignals commented on routes/projects/download.js:
```diff
@@ -3,70 +3,124 @@ const Json2csvTransform = require('json2csv').Transform;
+        .format('ESRI Shapefile')
+        .skipfailures()
+        .options(['-nln', 'projects']) // sets name of individual files in shapefile
+        .stream();
```
The shapefile creation should be pulled up into another function like you did with `getProjectsFeatureCollection` :)

chriswhong replied:
Done

allthesignals commented on routes/projects/index.js:
```diff
@@ -11,7 +11,7 @@ const getQueryFile = require('../../utils/get-query-file');
 
 const getProjectsFeatureCollection = async (app, query) => {
-  const SQL = buildProjectsSQL(query, 'shp');
+  const SQL = buildProjectsSQL(query, 'spatial');
```
GREAT, makes perfect sense. This should be documented in `buildProjectsSQL`.

allthesignals commented on routes/projects/download.js:
```diff
@@ -3,70 +3,124 @@ const Json2csvTransform = require('json2csv').Transform;
+
+      res.setHeader('Content-Type', 'text/csv');
+      // remove geom for csv streaming
+      app.db.stream(qs, (s) => {
```
In terms of extracting stuff into functions, I think this callback function here could be pulled up into a utility function above. We would pass `app.db.stream()` the utility function (instead of an anonymous func).

allthesignals commented on routes/projects/download.js:
```diff
@@ -3,70 +3,124 @@ const Json2csvTransform = require('json2csv').Transform;
+      const qs = new QueryStream(SQL);
+
+      const transformOpts = { highWaterMark: 16384, encoding: 'utf-8' };
+      const json2csv = new Json2csvTransform({}, transformOpts);
```
See comment below, these could be coupled inside the csv-handler utility function, no? This could be defined in the callback.

chriswhong commented on routes/projects/download.js:
```diff
@@ -3,70 +3,124 @@ const Json2csvTransform = require('json2csv').Transform;
-          'Duration in milliseconds:', data.duration,
-        );
+  // Set approrpiate download headers
+  res.setHeader('Content-disposition', `attachment; filename=projects.${filetype === 'shp' ? 'zip' : filetype}`);
```
great idea

julialucyhogan commented:
don't forget to rebase & squash before merging !!!

allthesignals commented:
This is really hard to setup for local development. Let's get some of these dependencies Dockerized so it's easier to setup fresh!

allthesignals commented:
Two things:

 - [ ] Need to fix running this in CI
 - [ ] Squash ur commits into one commit plz!

allthesignals commented:
@chriswhong blah, I need help: https://circleci.com/docs/2.0/building-docker-images/

julialucyhogan commented on .circleci/config.yml:
```diff
@@ -33,6 +31,12 @@ jobs:
+          command: cd ~ && mkdir .ssh && touch .ssh/known_hosts
       - run:
           name: Avoid hosts unknown for github
           command: ssh-keyscan zap-api.planninglabs.nyc >> ~/.ssh/known_hosts
```
can/should this just happen as one step? 
`command: mkdir -p ~/.ssh/known_hosts; zap-api.planninglabs.nyc >> ~/.ssh/known_hosts`
or maybe `&&` instead of the `;`, idk.

chriswhong replied:
I implemented this and amended the latest commit

chriswhong commented on routes/projects/download.js:
```diff
@@ -3,70 +3,128 @@ const Json2csvTransform = require('json2csv').Transform;
+
+  // build a query for the requested query params and filetype
+  // all three download types start with the same query (including geoms)
+  const SQL = buildProjectsSQL(query, 'download');
```
Good catch!  I refactored to pass `SQL` into `getProjectsFeatureCollection()`

Commit:
Add "prefiled" as public status option

Part of the CRM update was to add a prefiled optinon.  We addressed this
in PR #79, however we missed this place the dcp_publicstatus options
are listed.

allthesignals reviewed (approved):
OH! Thank you!

bfreeds commented:
Merging despite CI failing, as our CI currently is not working as expected and this will not break anything.
